### PR TITLE
OCPBUGS-24971: bump golang to 1.21

### PR DESCRIPTION
**Pull request by kevinrizza**
**Description of the change:**

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/operator-framework/operator-marketplace
 
-go 1.19
+go 1.21
```
Comment (bentito) [inline go.mod:3]: we were requiring go 1.19 but building on the 1.20 base image?

Comment (kevinrizza) [reply go.mod:3]: Apparently

File: .ci-operator.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
 build_root_image:
   name: release
   namespace: openshift
-  tag: rhel-8-release-golang-1.20-openshift-4.15
+  tag: rhel-8-release-golang-1.21-openshift-4.16
```
Comment (grokspawn) [inline .ci-operator.yaml:4]: Does this need to agree with https://github.com/operator-framework/operator-marketplace/pull/560/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R1 ?

File: Dockerfile [hunk 3]
```diff
@@ -1,15 +1,15 @@
-FROM registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.20-openshift-4.15 AS builder
+FROM registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.21-openshift-4.16 AS builder
 WORKDIR /go/src/github.com/operator-framework/operator-marketplace
 COPY . .
 RUN make osbs-build
 
-FROM registry.ci.openshift.org/ocp/4.15:base
+FROM registry.ci.openshift.org/ocp/4.16:base
```
Comment (grokspawn) [inline Dockerfile:6]:
```suggestion
FROM registry.ci.openshift.org/ocp/4.16:base-rhel9
```

**Pull request merged**